src/blob.jl

```diff
@@ -63,36 +63,40 @@ The number of bytes needed to allocate `T` itself.
 
 Defaults to `sizeof(T)`.
 """
-@generated function self_size(::Type{T}) where T
-    @assert isconcretetype(T)
+Base.@assume_effects :foldable function self_size(::Type{T}) where T
+    # This function is marked :total to encourage constant folding for this types-only
+    # static computation.
     if isempty(fieldnames(T))
-        quote
-            $(Expr(:meta, :inline))
-            $(sizeof(T))
-        end
+        sizeof(T)
     else
-        quote
-            $(Expr(:meta, :inline))
-            $(+(0, @splice i in 1:length(fieldnames(T)) begin
-                self_size(fieldtype(T, i))
-            end))
-        end
+        # Recursion is the fastest way to compile this, confirmed with benchmarks.
+        # Alternative considered: +(Iterators.map(self_size, fieldtypes(T))...)
+        # ~0.5ms for 5 fields, vs ~5ms for unrolling via splatting the fields.
+        # ~3ms for 20 fields, vs ~6ms for splatting.
+        # Note that splatting gives up after ~30 fields, whereas recursion remains robust.
+        _sum_field_sizes(T)
```

Comment:
It looks like it might be a good idea to add some kind of cutoff for the recursion to fall back to runtime computations for very large types?


That's what @aviatesk did here:
https://github.com/JuliaLang/julia/pull/54026/files#diff-12e7a6522633012a408b1bdee7639e8cb722617fe1a8ed6a3881bf4ad1ebdbbdR1369-R1370